d3/hyper.js

```diff
@@ -208,6 +231,61 @@ function generateDisplayNames(treeData) {
     }, common.allChildren);
 }
 
+// Function to find a node with a given operatorId assuming left-deep assignment
+function findNodeByOperatorId(node, operatorId) {
+    if (node.hasOwnProperty("properties") && node.properties.hasOwnProperty("operatorId") &&
+        node.properties.operatorId === operatorId) {
+        return node;
+    }
+    var children = common.allChildren(node);
+    if (children) {
+        var child;
+        for (var i = children.length; i-- > 0;) {
+            child = children[i];
+            if (i === 0) {
+                break;
+            }
+            if (child.hasOwnProperty("properties") && child.properties.hasOwnProperty("operatorId") &&
+                Number(child.properties.operatorId) <= Number(operatorId)) {
+                break;
+            }
+        }
+        return findNodeByOperatorId(child, operatorId);
+    }
+}
+
+// Function to add crosslinks between related nodes
+function addCrosslinks(root) {
+    var crosslinks = [];
+    var sourcenodes = [];
+    common.visit(root, function(node) {
+        switch (node.tag) {
+            case "explicitscan":
+                if (node.hasOwnProperty("properties") && node.properties.hasOwnProperty("source")) {
+                    sourcenodes.push({node: node, operatorId: node.properties.source});
+                }
+                break;
+            case "earlyprobe":
+                if (node.hasOwnProperty("properties") && node.properties.hasOwnProperty("builder")) {
+                    sourcenodes.push({node: node, operatorId: node.properties.builder});
+                }
+                break;
+            default:
+                if (node.hasOwnProperty("properties") && node.properties.hasOwnProperty("magic")) {
+                    sourcenodes.push({node: node, operatorId: node.properties.magic});
+                }
+                break;
+        }
+    }, common.allChildren);
+    // Add crosslinks from source to matching target node
+    sourcenodes.forEach(function(source) {
+        var targetnode = findNodeByOperatorId(root, source.operatorId);
```

Comment:
Instead of calling findNodeByOperatorId multiple times and thereby traversing the tree multiple times, I would be in favor of creating a hash table for this. Something along the lines of:

```javascript
var operatorsById = {};
common.visit(root, function(node) {
   if (node.hasOwnProperty("properties") && node.properties.hasOwnProperty("operatorId")) {
      operatorsById[node.properties.operatorId] = node;
   }
}
```

and then this line here simply becomes
```javascript
var targetNode = operatorsById[source.operatorId];
```

we could even fill the `operatorsById` map as part of the tree traversal started in line 261 and wouldn't even need an additional traversal.

Reply:
Good suggestion that also facilitates crosslinks for earlier plans that did not have an operatorId defined for every node.